In [109]:
import numpy as np 
import pandas as pd

In [110]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv', 'USA.csv', 'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                'USA1.csv', 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv', 
                                 'Italy.csv', 'South_Korea.csv', 'Nigeria.csv',
                                'Colombia.csv', 'Paraguay.csv', 'Russia.csv',
                                'Algeria.csv', 'Chile.csv', 'Yugoslavia.csv',
                                'Scotland.csv', 'Romania.csv', 'Greece.csv',
                                'Honduras.csv', 'Austria.csv', 'Bulgaria.csv', 
                                 'IvoryCoast.csv', 'SouthAfrica.csv', 'RepIreland.csv',
                                'Slovenia.csv', 'Czechoslovakia.csv', 'Hungary.csv', 
                                'Egypt.csv', 'Norway.csv', 'Peru.csv']))



In [111]:
# 158 with wc 2022 teams

In [112]:
df.shape

(249, 17)

In [113]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [114]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.626423   0.121791      0.361573  0.201254   
xg_z             0.626423  1.000000   0.403630      0.658816  0.317344   
crosses_z        0.121791  0.403630   1.000000      0.646816  0.261067   
boxtouches_z     0.361573  0.658816   0.646816      1.000000  0.370556   
passes_z         0.201254  0.317344   0.261067      0.370556  1.000000   
progpasses_z     0.057748  0.110861   0.283617      0.241600  0.743343   
takeons_z        0.280116  0.370343   0.144843      0.243476  0.390093   
progruns_z       0.210852  0.376921   0.271905      0.382399  0.578904   
tackles_z        0.220027  0.345147   0.224616      0.386077  0.736886   
interceptions_z  0.185047  0.274207   0.229022      0.284006  0.748676   
clearances_z    -0.206795 -0.393340  -0.265786     -0.421671 -0.119142   
blocks_z         0.044809  0.052933   0.032907      0.123944  0.381609   
aerials_z        0.025145  0.019199   0.229716      0.223041 -0.019652   
fouls_z         -0.080193 -0.010628   0.081504      0.045067  0.080286   
fouled_z         0.062243  0.172793   0.174428      0.175380  0.307298   
nsxg_z           0.389946  0.648870   0.692136      0.858338  0.481129   
results         -0.603331 -0.314959  -0.012455     -0.205333 -0.145072   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.057748   0.280116    0.210852   0.220027   
xg_z                 0.110861   0.370343    0.376921   0.345147   
crosses_z            0.283617   0.144843    0.271905   0.224616   
boxtouches_z         0.241600   0.243476    0.382399   0.386077   
passes_z             0.743343   0.390093    0.578904   0.736886   
progpasses_z         1.000000   0.150204    0.314679   0.498668   
takeons_z            0.150204   1.000000    0.550012   0.387286   
progruns_z           0.314679   0.550012    1.000000   0.512019   
tackles_z            0.498668   0.387286    0.512019   1.000000   
interceptions_z      0.559612   0.381535    0.416062   0.695414   
clearances_z        -0.006009  -0.116678   -0.093093  -0.207252   
blocks_z             0.343120   0.167480    0.272149   0.266387   
aerials_z            0.032038  -0.132021   -0.097711   0.013539   
fouls_z              0.221248   0.125744    0.035097   0.020043   
fouled_z             0.190707   0.341139    0.251172   0.300686   
nsxg_z               0.364531   0.392535    0.468336   0.461608   
results             -0.026958  -0.179808   -0.198999  -0.174490   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.185047     -0.206795  0.044809   0.025145 -0.080193   
xg_z                    0.274207     -0.393340  0.052933   0.019199 -0.010628   
crosses_z               0.229022     -0.265786  0.032907   0.229716  0.081504   
boxtouches_z            0.284006     -0.421671  0.123944   0.223041  0.045067   
passes_z                0.748676     -0.119142  0.381609  -0.019652  0.080286   
progpasses_z            0.559612     -0.006009  0.343120   0.032038  0.221248   
takeons_z               0.381535     -0.116678  0.167480  -0.132021  0.125744   
progruns_z              0.416062     -0.093093  0.272149  -0.097711  0.035097   
tackles_z               0.695414     -0.207252  0.266387   0.013539  0.020043   
interceptions_z         1.000000     -0.117102  0.311170  -0.009078  0.145117   
clearances_z           -0.117102      1.000000  0.202925   0.285981  0.204690   
blocks_z                0.311170      0.202925  1.000000   0.023616  0.187962   
aerials_z              -0.009078      0.285981  0.023616   1.000000  0.127747   
fouls_z                 0.145117      0.204690  0.187962   0.127747  1.000000   
fouled_z                0.320314      0.034775  0.079754   0.008107  0.283676   
nsxg_z                  0.405356     -0.413704  0.154483   0.151427  0.096396   
results                -0.109239      0.107403 -0.063775  -0.034183  0.095687   

                 fo

In [115]:
df.shape

(249, 17)

In [116]:
df.duplicated().value_counts()

False    249
dtype: int64

In [117]:
df.duplicated().value_counts()

False    249
dtype: int64

# convert to csv

In [118]:
df.to_csv('WC-stats.csv', index=False)


In [119]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [120]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [121]:
y = df['results']

In [122]:
X = df.drop(columns=['results'])

In [123]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [124]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.39545071246627417

In [125]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [126]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 0.808 +/- 0.079
    xg_z    0.063 +/- 0.017
    nsxg_z  0.055 +/- 0.018
    crosses_z0.043 +/- 0.014
    progruns_z0.041 +/- 0.013
    fouled_z0.026 +/- 0.009
neg_mean_absolute_percentage_error
    goals_z 0.606 +/- 0.069
    nsxg_z  0.049 +/- 0.021
    progruns_z0.036 +/- 0.015
    fouled_z0.033 +/- 0.014
    tackles_z0.030 +/- 0.010
    fouls_z 0.029 +/- 0.011
neg_mean_squared_error
    goals_z 60.654 +/- 5.963
    xg_z    4.748 +/- 1.245
    nsxg_z  4.135 +/- 1.349
    crosses_z3.201 +/- 1.050
    progruns_z3.082 +/- 0.962
    fouled_z1.928 +/- 0.687


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.